pkg/sql/colexec/auto_incr.go

```diff
@@ -0,0 +1,400 @@
+// Copyright 2022 Matrix Origin
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package colexec
+
+import (
+	"context"
+	"fmt"
+	"math"
+
+	"github.com/matrixorigin/matrixone/pkg/container/batch"
+	"github.com/matrixorigin/matrixone/pkg/container/nulls"
+	"github.com/matrixorigin/matrixone/pkg/container/types"
+	"github.com/matrixorigin/matrixone/pkg/container/vector"
+	"github.com/matrixorigin/matrixone/pkg/pb/plan"
+	"github.com/matrixorigin/matrixone/pkg/sql/errors"
+	"github.com/matrixorigin/matrixone/pkg/vm/engine"
+	"github.com/matrixorigin/matrixone/pkg/vm/engine/tae/moengine"
+	"github.com/matrixorigin/matrixone/pkg/vm/process"
+)
+
+var AUTO_INCR_TABLE = "%!%mo_increment_columns"
+var AUTO_INCR_TABLE_COLNAME []string = []string{"name", "offset", "step"}
+
+type AutoIncrParam struct {
+	eg      engine.Engine
+	db      engine.Database
+	rel     engine.Relation
+	ctx     context.Context
+	proc    *process.Process
+	colDefs []*plan.ColDef
+}
+
+func UpdateInsertBatch(e engine.Engine, db engine.Database, ctx context.Context, proc *process.Process, ColDefs []*plan.ColDef, bat *batch.Batch, tableID string) error {
+	incrParam := &AutoIncrParam{
+		eg:      e,
+		db:      db,
+		ctx:     ctx,
+		proc:    proc,
+		colDefs: ColDefs,
+	}
+
+	offset, step, err := getRangeFromAutoIncrTable(incrParam, bat, tableID)
+	if err != nil {
+		return err
+	}
+
+	if err = updateBatchImpl(ColDefs, bat, offset, step); err != nil {
+		return err
+	}
+	return nil
+}
+
+func UpdateInsertValueBatch(e engine.Engine, ctx context.Context, proc *process.Process, p *plan.InsertValues, bat *batch.Batch) error {
+	ColDefs := p.ExplicitCols
+	orderColDefs(p.OrderAttrs, ColDefs)
+	db, err := e.Database(ctx, p.DbName, proc.TxnOperator)
+	if err != nil {
+		return err
+	}
+	rel, err := db.Relation(ctx, p.TblName)
+	if err != nil {
+		return err
+	}
+	return UpdateInsertBatch(e, db, ctx, proc, ColDefs, bat, rel.GetTableID(ctx))
+}
+
+func getRangeFromAutoIncrTable(param *AutoIncrParam, bat *batch.Batch, tableID string) ([]int64, []int64, error) {
+	offset, step := make([]int64, 0), make([]int64, 0)
+	var err error
+	for i, col := range param.colDefs {
+		if !col.AutoIncrement {
+			continue
+		}
+		var d, s int64
+		param.rel, err = param.db.Relation(param.ctx, AUTO_INCR_TABLE)
+		if err != nil {
+			return nil, nil, err
+		}
+		if d, s, err = getOneColRangeFromAutoIncrTable(param, bat, tableID+"_"+col.Name, i); err != nil {
```

Comment:
Actually, figure out all auto increment columns at first. Then get the range of these columns from auto_incr_table in only one transaction. It is unnecessary to start multiple transactions here.